ldap/servers/plugins/replication/cl5_api.c

```diff
@@ -2192,13 +2196,33 @@ _cl5TrimMain(void *param)
     struct timespec prev_time = {0};
     Replica *replica = (Replica *)param;
     cldb_Handle *cldb = replica_get_cl_info(replica);
-    int32_t trimInterval = cldb->clConf.trimInterval;
+    int32_t trimInterval;
+
+    if (cldb == NULL) {
+        /* This can happened in race condition
+         * when the cldb_SetReplicaDB is called but the
+         * dispatching of_cl5TrimMain thread is slow.
+         * So trimming could be have been stopped (ruv reload) that
+         * clears the cldb
+         */
+        return 0;
+    }
+    trimInterval = cldb->clConf.trimInterval;
 
     /* Get the initial current time for checking the trim interval */
     clock_gettime(CLOCK_MONOTONIC, &prev_time);
 
     /* Lock the CL state, and bump the thread count */
     pthread_mutex_lock(&(cldb->stLock));
+
+    /* First check that no other trimming thread is running */
+    if (cldb->trimmingOnGoing) {
+        pthread_mutex_unlock(&(cldb->stLock));
+        return 0;
+    }
+
+    /* Now trimming thread can start */
+    cldb->trimmingOnGoing = 1;
```

Comment:
I am not sure if it's a legit concern but do we also need to take into an account other places where we call `slapi_counter_increment(cldb->clThreads);` and `slapi_counter_decrement(cldb->clThreads);`? Like cl5CreateReplayIteratorEx(), cl5CreateReplayIterator(), etc.

Reply:
Good point !

One part of the patch is prevent the creation of several trimming threads. It uses a cldb flag (trimmingOnGoing) protected by stLock. This part does not really or change the clThread mechanism (see below)

A second part of the patch (test cldb != NULL) is to address a rare race condition (depending on thread dispatching) that can crash the server.
clThreads is a way to prevent to close/remove cldb under others threads (remove changelog file, read clRUV, export/import, write update, trimming/RA cursor, ...). This mechanism is not perfect and cldb could be cleared from the replica while _cl5TrimMain is called. We could improve clThreads, for example moving it from cldb to the replica, but I found this solution a bit overkill compare to the test cldb != NULL. Do you prefer to remove the test and change the clThreads mechanism ?

Reply:
Okay, thank you for the explanations! I had concerns regarding the trimmingOnGoing flag, actually, and you explained it. :)
I agree that (test cldb != NULL) is the way to go for the second issue.